rir/src/compiler/analysis/visibility.cpp

```diff
@@ -4,30 +4,26 @@
 namespace rir {
 namespace pir {
 
-AbstractResult VisibilityAnalysis::apply(CurrentVisibility& vis,
+AbstractResult VisibilityAnalysis::apply(LastVisibilityUpdate& vis,
                                          Instruction* i) const {
-    // Default: visible
-    if (!code->entry->isEmpty() && i == *code->entry->begin())
-        vis.state = CurrentVisibility::Visible;
-
+    AbstractResult res;
+    auto isVisibilityChanging = [&]() {
+        if (vis.last != i) {
+            vis.last = i;
+            vis.observable.clear();
+            res.update();
+        }
+    };
     switch (i->tag) {
     case Tag::Invisible:
-        if (vis.state != CurrentVisibility::Invisible) {
-            vis.state = CurrentVisibility::Invisible;
-            return AbstractResult::Updated;
-        }
-        break;
     case Tag::Visible:
     case Tag::LdVar:
     case Tag::LdVarSuper:
     case Tag::Extract1_1D:
     case Tag::Extract2_1D:
     case Tag::Extract1_2D:
     case Tag::Extract2_2D:
-        if (vis.state != CurrentVisibility::Visible) {
-            vis.state = CurrentVisibility::Visible;
-            return AbstractResult::Updated;
-        }
+        isVisibilityChanging();
```

Comment:
Are these instructions the only ones that change visibility? Could you use `Instruction::mightChangeVisibility` here?

Reply:
unfortunately not. I can only be sure that an instruction is that last visible change to the visibility flag if I am certain that it changes visibility. the mightChange query doesn't help. i don't see an easy way to get does certainly change visibility at the instruction level...